File: apps/web/src/features/catalog/notation.ts
/**
 * Engineering notation, the browser's half of it.
 *
 * The API owns the stored value: it parses what was typed with `Decimal`, exactly, and
 * answers with both the value and its display form. This module exists so the field can
 * show `4.7k` under `4k7` while it is being typed, before anything is sent, and so an
 * unreadable number is caught without a round trip (requirements 7.3, 7.4). It follows the
 * same rules as `catalog/domain/notation.py`; a double is good enough for a preview at four
 * significant digits, and is never what gets stored.
 */

// Case matters: m is milli, M is mega. R is the resistor convention for "no prefix" (1R5
// is 1.5) and u the ASCII spelling of µ. K is absent on purpose: k is kilo, K is kelvin.
const EXPONENT_BY_PREFIX: Record<string, number> = {
  p: -12,
  n: -9,
  µ: -6, // U+00B5 MICRO SIGN, which NFKC folds into the Greek mu below
  μ: -6,
  u: -6,
  m: -3,
  R: 0,
  k: 3,
  M: 6,
  G: 9,
  T: 12,
};

const PREFIX_BY_EXPONENT: Record<number, string> = {
  "-12": "p",
  "-9": "n",
  "-6": "µ",
  "-3": "m",
  0: "",
  3: "k",
  6: "M",
  9: "G",
  12: "T",
};

const MIN_PREFIX_EXPONENT = -12;
const MAX_PREFIX_EXPONENT = 12;
const SIGNIFICANT_DIGITS = 4;

const PREFIXES = Object.keys(EXPONENT_BY_PREFIX).join("");
const DECIMAL = String.raw`[+-]?(?:\d+(?:\.\d+)?|\.\d+)`;
// 4700, 4.7e3
const PLAIN = new RegExp(`^${DECIMAL}(?:[eE][+-]?\\d+)?$`);
// 10k, 100n, 2.2µ
const SUFFIXED = new RegExp(`^(${DECIMAL})([${PREFIXES}])$`);
// 4k7, 2u2, 1R5: the prefix stands in for the decimal point
const INFIXED = new RegExp(`^([+-]?\\d+)([${PREFIXES}])(\\d+)$`);

/**
 * What was typed, in SI base units, or null when it can't be read. A trailing unit is
 * accepted only when it is the attribute's own, so `100nF` is a farad and `100nH` is not.
 */
export function parseSi(text: string, unit?: string | null): number | null {
  const cleaned = text.normalize("NFKC").trim();
  // The prefix reading goes first, so "10m" stays milli even on a metre attribute.
  const direct = read(cleaned);
  if (direct !== null) return direct;
  const bare = withoutUnit(cleaned, unit);
  return bare === null ? null : read(bare);
}

/** The value with the prefix that puts the mantissa in [1, 1000), four digits at most. */
export function formatSi(value: number, unit?: string | null): string {
  const symbol = unit ?? "";
  if (!Number.isFinite(value)) return "";
  if (value === 0) return `0${symbol}`;
  const exponent = engineeringExponent(value);
  const mantissa = Number((value / 10 ** exponent).toPrecision(SIGNIFICANT_DIGITS));
  return `${mantissa}${PREFIX_BY_EXPONENT[exponent] ?? ""}${symbol}`;
}

function read(text: string): number | null {
  if (PLAIN.test(text)) return finite(Number(text));
  const suffixed = SUFFIXED.exec(text);
  if (suffixed) return scaled(suffixed[1], suffixed[2]);
  const infixed = INFIXED.exec(text);
  if (infixed) return scaled(`${infixed[1]}.${infixed[3]}`, infixed[2]);
  return null;
}

/** Scaling through the exponent notation rather than a multiplication, which would drift. */
function scaled(mantissa: string | undefined, prefix: string | undefined): number | null {
  const exponent = prefix === undefined ? undefined : EXPONENT_BY_PREFIX[prefix];
  if (mantissa === undefined || exponent === undefined) return null;
  return finite(Number(`${mantissa}e${exponent}`));
}

function withoutUnit(text: string, unit: string | null | undefined): string | null {
  if (!unit) return null;
  const symbol = unit.normalize("NFKC");
  if (!text.endsWith(symbol) || text.length === symbol.length) return null;
  return text.slice(0, -symbol.length);
}

function finite(value: number): number | null {
  return Number.isFinite(value) ? value : null;
}

/**
 * The multiple of three to divide by. Rounding to four digits happens first, so 999.99
 * becomes 1k rather than 1000.
 */
function engineeringExponent(value: number): number {
  const rounded = Math.abs(value).toExponential(SIGNIFICANT_DIGITS - 1);
  const decades = Number(rounded.slice(rounded.indexOf("e") + 1));
  const multipleOfThree = Math.floor(decades / 3) * 3;
  return Math.min(MAX_PREFIX_EXPONENT, Math.max(MIN_PREFIX_EXPONENT, multipleOfThree));
}

File: apps/api/src/wiredex/catalog/domain/notation.py
"""Engineering notation: the only way a number enters the catalog domain.

Pure functions over Decimal, because two spellings of the same magnitude have to compare
exactly equal (4k7 == 4700) and float would make 100n land next to, not on, 1e-7.
"""

import re
import unicodedata
from decimal import ROUND_HALF_UP, Decimal

from wiredex.catalog.domain.errors import InvalidNumberError
from wiredex.catalog.domain.values import SiValue, Unit

# Case matters: m is milli and M is mega. R is the resistor convention for "no prefix"
# (1R5 is 1.5) and u is the ASCII spelling of µ. K is absent on purpose: k is kilo,
# K is kelvin, and guessing which one was meant is how wrong values get stored.
_EXPONENT_BY_PREFIX = {
    "p": -12,
    "n": -9,
    "µ": -6,  # U+00B5 MICRO SIGN, what format_si prints
    "μ": -6,  # U+03BC GREEK SMALL LETTER MU, what phones type and NFKC turns µ into
    "u": -6,
    "m": -3,
    "R": 0,
    "k": 3,
    "M": 6,
    "G": 9,
    "T": 12,
}
# What format_si prints back: one symbol per exponent, µ over u, nothing over R.
_PREFIX_BY_EXPONENT = {-12: "p", -9: "n", -6: "µ", -3: "m", 0: "", 3: "k", 6: "M", 9: "G", 12: "T"}
MIN_PREFIX_EXPONENT = -12
MAX_PREFIX_EXPONENT = 12
SIGNIFICANT_DIGITS = 4

_PREFIXES = "".join(_EXPONENT_BY_PREFIX)
_DECIMAL = r"[+-]?(?:\d+(?:\.\d+)?|\.\d+)"
# 4700, 4.7e3
_PLAIN = re.compile(rf"^{_DECIMAL}(?:[eE][+-]?\d+)?$")
# 10k, 100n, 2.2µ
_SUFFIXED = re.compile(rf"^({_DECIMAL})([{_PREFIXES}])$")
# 4k7, 2u2, 1R5: the prefix stands in for the decimal point
_INFIXED = re.compile(rf"^([+-]?\d+)([{_PREFIXES}])(\d+)$")
# Only to explain the rejection, never to accept it.
_KELVIN = re.compile(rf"^{_DECIMAL}K\d*$")

_EXAMPLES = "write it like 4k7, 4700 or 4.7e3"


def parse_si(text: str, unit: Unit | None = None) -> SiValue:
    """Reads engineering notation into SI base units, rejecting anything ambiguous."""
    cleaned = normalize_symbols(text).strip()
    number = _read(cleaned)
    if number is None:
        # The prefix reading goes first, so "10m" stays milli even on a metre attribute;
        # a trailing unit is only dropped when it is the attribute's own.
        without_unit = _drop_unit(cleaned, unit)
        number = None if without_unit is None else _read(without_unit)
    if number is None:
        raise InvalidNumberError(_explain(text, unit))
    return SiValue(number.normalize())


def normalize_symbols(text: str) -> str:
    """One spelling for look-alikes: µ and μ, the ohm sign and omega, full-width digits."""
    return unicodedata.normalize("NFKC", text)


def format_si(value: SiValue, unit: Unit | None = None) -> str:
    """Prints a value with the prefix that puts the mantissa in [1, 1000), four digits at most."""
    mantissa, exponent = _engineering(value.value)
    symbol = "" if unit is None else str(unit)
    return f"{mantissa.normalize():f}{_PREFIX_BY_EXPONENT[exponent]}{symbol}"


def _read(text: str) -> Decimal | None:
    if _PLAIN.match(text):
        return Decimal(text)
    suffixed = _SUFFIXED.match(text)
    if suffixed is not None:
        return Decimal(suffixed[1]).scaleb(_EXPONENT_BY_PREFIX[suffixed[2]])
    infixed = _INFIXED.match(text)
    if infixed is not None:
        return Decimal(f"{infixed[1]}.{infixed[3]}").scaleb(_EXPONENT_BY_PREFIX[infixed[2]])
    return None


def _drop_unit(text: str, unit: Unit | None) -> str | None:
    if unit is None:
        return None
    symbol = str(unit)
    if not text.endswith(symbol) or len(text) == len(symbol):
        return None
    return text[: -len(symbol)]


def _explain(text: str, unit: Unit | None) -> str:
    if _KELVIN.match(text.strip()):
        return f"{text!r}: K is kelvin, k is kilo — {_EXAMPLES}"
    if unit is not None:
        return f"{text!r} is not a number in {unit} — {_EXAMPLES}"
    return f"{text!r} is not a number — {_EXAMPLES}"


def _engineering(number: Decimal) -> tuple[Decimal, int]:
    if not number:
        return Decimal(0), 0
    exponent = _prefix_exponent(number)
    mantissa = _to_significant(number.scaleb(-exponent))
    if abs(mantissa) >= 1000 and exponent < MAX_PREFIX_EXPONENT:  # noqa: PLR2004  the [1, 1000) rule
        # Rounding pushed the mantissa out of range, as 999.99 does: step up a prefix.
        exponent += 3
        mantissa = _to_significant(number.scaleb(-exponent))
    return mantissa, exponent


def _prefix_exponent(number: Decimal) -> int:
    # Outside p..T the mantissa leaves [1, 1000) rather than inventing a prefix nobody reads.
    multiple_of_three = number.adjusted() // 3 * 3
    return max(MIN_PREFIX_EXPONENT, min(MAX_PREFIX_EXPONENT, multiple_of_three))


def _to_significant(mantissa: Decimal) -> Decimal:
    step = Decimal(1).scaleb(mantissa.adjusted() - (SIGNIFICANT_DIGITS - 1))
    return mantissa.quantize(step, rounding=ROUND_HALF_UP)
